src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/ErrorableExtensions.cs

```diff
@@ -9,142 +9,193 @@ namespace Microsoft.Azure.Batch.SoftwareEntitlement.Common
     public static class ErrorableExtensions
     {
         /// <summary>
-        /// Combine two <see cref="Errorable{T}"/> values
+        /// Combine two <see cref="Errorable{T}"/> values into a single value containing a tuple
         /// </summary>
-        /// <typeparam name="T">Type of value possibly present in <paramref name="first"/>.</typeparam>
-        /// <typeparam name="A">Type of value possibly present in <paramref name="second"/>.</typeparam>
-        /// <param name="first">First errorable value.</param>
-        /// <param name="second">Second errorable value.</param>
-        /// <param name="whenSuccessful">Action to take when both values are available.</param>
-        /// <param name="whenFailure">Action to take if either (or both) errorables have errors.</param>
-        public static void Combine<T, A>(
-            this Errorable<T> first,
-            Errorable<A> second,
-            Action<T, A> whenSuccessful,
-            Action<IEnumerable<string>> whenFailure)
+        /// <remarks>
+        /// Works as a logical <c>AND</c> - if both inputs are successful, the output is successful; 
+        /// if either input is a failure, the output is a failure. All error messages are preserved.
+        /// </remarks>
+        /// <typeparam name="A">Type of value held by <paramref name="left"/>.</typeparam>
+        /// <typeparam name="B">Type of value held by <paramref name="right"/>.</typeparam>
+        /// <param name="left">First <see cref="Errorable{T}"/> to combine.</param>
+        /// <param name="right">Second <see cref="Errorable{T}"/> to combine.</param>
+        /// <returns>An <see cref="Errorable{T}"/> containing either both values or a combined 
+        /// set of error messages.</returns>
+        public static Errorable<(A, B)> And<A, B>(this Errorable<A> left, Errorable<B> right)
         {
-            if (first == null)
+            if (left == null)
             {
-                throw new ArgumentNullException(nameof(first));
+                throw new ArgumentNullException(nameof(left));
             }
 
-            if (second == null)
+            if (right == null)
             {
-                throw new ArgumentNullException(nameof(second));
+                throw new ArgumentNullException(nameof(right));
             }
 
-            if (whenSuccessful == null)
-            {
-                throw new ArgumentNullException(nameof(whenSuccessful));
-            }
+            return left.Match(
+                whenSuccessful: leftValue => right.Match(
+                    whenSuccessful: rightValue => Errorable.Success((leftValue, rightValue)),
+                    whenFailure: errors => Errorable.Failure<(A, B)>(errors)),
+                whenFailure: leftErrors => right.Match(
+                    whenSuccessful: rightValue => Errorable.Failure<(A, B)>(leftErrors),
```

Comment:
>rightValue [](start = 36, length = 10)

what do you feel about using `_` for "discarded" arguments? it would make it clear that `rightValue` is not used when `left` has failed and vice versa

Reply:
actually not vice versa because you have to use `leftValue` in the other branch :)

---
In reply to: [149828355](https://github.com/Azure/azure-batch-software-entitlement/pull/74#discussion_r149828355) [](ancestors = 149828355)

Reply:
Yeah, that makes sense - happy to do that.